File: 0020-valid-parentheses/0020-valid-parentheses.cpp
// TC- O(n)
// SC- O(n)

class Solution {
public:
    bool isValid(string s) {
        stack<char> st;

        for(int i=0; i<s.size(); i++){
            char ch = s[i];
            if(ch=='{' or ch=='(' or ch=='[')
                st.push(ch);
            else if(ch=='}' or ch==')' or ch==']'){
                if(st.empty())
                    return false;
                char top = st.top();
                st.pop();
                if((ch==')' and top!='(') or (ch==']' and top!='[') or (ch=='}' and top!='{'))
                    return false;
            }
        }

        if(!st.empty())
            return false;
        
        return true;
    }
};

File: 0020-valid-parentheses/0020-valid-parentheses.py
# TC- O(n)
# SC- O(n)

class Solution:
    def isValid(self, s: str) -> bool:
        self.stack = []

        for ch in s:
            if ch in ["(", "[", '{']:
                self.stack.append(ch)
            else:
                if not self.stack:
                    return False
                popped = self.stack.pop()
                if (ch==")" and popped is not "(") or (ch=="]" and popped is not "[") or (ch=="}" and popped is not "{"):
                    return False
        if self.stack:
            return False
        return True